Header pre-release fixes

paulrobertlloyd opened the pull request:
## Description

This PR addresses a few issues that have been picked up as we prepare to release the updated header component.

1. **Implements link styling refactoring** from #1419, without implementing any stylistic changes for hover behaviour.
   - Well, except that the hover colour for links on a white background is now dark blue, going to a slighter darker shade on active, mimicking the colours used for the search submit button.
   - This refactor also fixes an issue with header print styles, reported in #1434

2. **Fixes the display of different header colour combinations** by removing the toggling of the top margin and border based on screen size and colour variant, reported in #561 (https://github.com/nhsuk/nhsuk-frontend/issues/561#issuecomment-3000949245).
  - Any combination of blue/white header and blue/white navigation will appear with the same spacing and metrics, with only border and background colours changing as needed. 
  - Also refactors the navigation styles, ensuring borders in the overflow menu line up with the top border for the navigation container, and that a border is shown if there is both a white navigation and white overflow menu.
  - We don’t have tests for every variant of the header colours; should we add them?

3. **Adds a maximum character length to the service name**. This is set `40ch`, ensuring super long names wrap earlier (and look a bit more visually balanced). See https://github.com/nhsuk/nhsuk-frontend/issues/997#issuecomment-3001140039

## Checklist

- [ ] Tested against our [testing policy](https://github.com/nhsuk/nhsuk-frontend/blob/master/docs/contributing/testing.md) (Resolution, Browser & Accessibility)
- [ ] Follows our [coding standards and style guide](https://github.com/nhsuk/nhsuk-frontend/blob/master/docs/contributing/coding-standards.md)
- [ ] CHANGELOG entry

frankieroberto commented:
@paulrobertlloyd looks like the mobile Header layout has changed to retain the horizontal line above the nav, was that deliberate?

Agree it’s probably worth having examples of all 4 header blue/white combinations in the review app, if only to stop us forgetting about them in future.  Might also be worth having some examples of longer service names to be able to easily test the wrapping (which looks good!)?

paulrobertlloyd commented:
> looks like the mobile Header layout has changed to retain the horizontal line above the nav, was that deliberate?

It was, part of my long running ‘make this component less complicated’ campaign. While it does add some vertical height, it keeps the design consistent across breakpoints, and stops having to juggle margins not just across breakpoints but also the 4 (!) colour combinations, making everything just a bit more predictable.

anandamaryon1 commented:
Thanks @paulrobertlloyd this is helpful. 

Two tiny quibbles:

* Could we up the max characters in service name from `40ch`, to something link `54ch` to closer approximate a standard line reading length? 
* I know there's some logic to it and it makes sense in the code but I feel adding the border between the nav and icon or search on the all blue or all white header on mobile is a regression design-wise. With above-the-fold space on mobile at such a premium I feel any pixels we can shave do help quite a lot. – does adding that back in that ruin your refactor efforts?

anandamaryon1 commented:
@colinrotherham Very happy, looking great!

paulrobertlloyd commented:
Have pushed a few commits which:

- Moves and renames modifier to make the navigation white so that it should now be added to navigation.class (`nhsuk-header__navigation--white`). **This will need a change log entry and updating the service manual**.
- Uses a `box-shadow` instead of `border-top` so that the navigation maintains the same height regardless of colour and no gap appears between a focused item and the top part of the header
- Reinstates the previous layout behaviour where, in instances where the header and navigation have the same colour, the border disappears and a negative margin is added until the tablet breakpoint. Consolidated and added comments around this behaviour.
- Updates the white header such that only the organisation name is black, all other components take on the link colour. I tried to get the service name to remain black, but given the different combinations of service links, it’s pretty hard to get this to work consistently – AND – this isn’t something that is really supported anyway (the white header being for organisations).
- Updated the reference images
